core/src/main/java/com/microsoft/applicationinsights/internal/heartbeat/HeartBeatProvider.java

```diff
@@ -0,0 +1,304 @@
+      try {
+
+        if (!heartbeatProperties.containsKey(propertyName)) {
+          InternalLogger.INSTANCE.trace("The property %s is not already present. It will be added", propertyName);
+        }
+        if (HeartbeatDefaultPayload.isDefaultKeyword(propertyName)) {
+          setResult = false;
+          throw new Exception("heartbeat beat property specified is a reserved property");
+        }
+
+        HeartBeatPropertyPayload payload = new HeartBeatPropertyPayload();
+        payload.setHealthy(isHealthy);
+        payload.setPayloadValue(propertyValue);
+        heartbeatProperties.put(propertyName, payload);
+        setResult = true;
+
+      }
+      catch (Exception e) {
+        InternalLogger.INSTANCE.warn("failed to set heartbeat property name %s, value %s, "
+            + "stack trace is: %s", propertyName, propertyValue, ExceptionUtils.getStackTrace(e));
+      }
+    }
+    else {
+      InternalLogger.INSTANCE.warn("cannot set property without property name");
+    }
+    return setResult;
+  }
+
+  @Override
+  public boolean isHeartBeatEnabled() {
+    return isEnabled;
+  }
+
+  @Override
+  public void setHeartBeatEnabled(boolean isEnabled) {
+    this.isEnabled = isEnabled;
+  }
+
+  @Override
+  public List<String> getExcludedHeartBeatPropertyProviders() {
+    return this.disabledHeartBeatPropertiesProviders;
+  }
+
+  @Override
+  public void setExcludedHeartBeatPropertyProviders(
+      List<String> excludedHeartBeatPropertyProviders) {
+    this.disabledHeartBeatPropertiesProviders = excludedHeartBeatPropertyProviders;
+  }
+
+  @Override
+  public long getHeartBeatInterval() {
+    return this.interval;
+  }
+
+  @Override
+  public void setHeartBeatInterval(long timeUnit) {
+    // user set time unit in seconds
+    if (timeUnit <= HeartBeatProviderInterface.MINIMUM_HEARTBEAT_INTERVAL) {
+      this.interval = HeartBeatProviderInterface.MINIMUM_HEARTBEAT_INTERVAL;
+    }
+    else {
+      this.interval = timeUnit;
+    }
+  }
+
+  @Override
+  public List<String> getExcludedHeartBeatProperties() {
+    return this.disableDefaultProperties;
+  }
+
+  @Override
+  public void setExcludedHeartBeatProperties(List<String> excludedHeartBeatProperties) {
+    this.disableDefaultProperties = excludedHeartBeatProperties;
+  }
+
+  @Override
+  public void stop(long timeout, TimeUnit timeUnit) {
+    propertyUpdateService.shutdown();
+    heartBeatSenderService.shutdown();
+    try {
+      propertyUpdateService.awaitTermination(timeout, timeUnit);
+      heartBeatSenderService.awaitTermination(timeout, timeUnit);
+    }
+    catch (InterruptedException e) {
+      InternalLogger.INSTANCE.warn("unable to successfully terminate heartbeat module, "
+          + "encountered and exception with stacktrace, %s", ExceptionUtils.getStackTrace(e));
+    }
+  }
+
+  /**
+   * Send the heartbeat item synchronously to application insights backend.
+   */
+  private void send() {
+
+    synchronized (lock) {
```

Comment:
What is the lock guarding? The ConcurrentMap keeps the CRUD to it thread-safe. I think you can just remove the lock and keep the atomic int

Reply:
@littleaj my intention here is to send heartBeats synchronously.  Telemetry client's track() method sends Telemetry asynchronously which is not desired here. I wanted to guard against an untoward situation like this. I have removed the Atomic version to primitive as you mentioned we are already locking and therefore avoid overhead.